src/ar_http_iface_cowboy_handler.erl

```diff
@@ -659,21 +665,139 @@ do_handle(Method, [<<"height">>], Req)
-		<<>> ->
-			do_handle(<<"GET">>, [<<"tx">>, Hash, <<"data.html">>], Req);
-		_ ->
-			not_found(Req)
+handle(<<"GET">>, [<<Hash:43/binary, MaybeExt/binary>>], Req) ->
+	case ar_meta_db:get(apex_domain) of
+		not_found ->
+			handle(<<"GET">>, [<<"tx">>, Hash, <<"data.", MaybeExt/binary>>], Req);
+		ApexDomain ->
+			handle_serve_tx_1(ApexDomain, Hash, MaybeExt, Req)
 	end;
 
 %% @doc Catch case for requests made to unknown endpoints.
 %% Returns error code 400 - Request type not found.
-do_handle(_, _, Req) ->
+handle(_, _, Req) ->
+	not_found(Req).
+
+handle_serve_root(ApexDomain, Req) ->
+	Domain = cowboy_req:host(Req),
+	case get_domain_labeling(ApexDomain, Domain) of
+		apex -> return_info(Req);
+		{labeled, _} -> redirect_to_apex(ApexDomain, Req);
+		{custom, Domain} -> resolve_custom_domain_tx(Domain, Req)
+	end.
+
+handle_serve_tx_1(ApexDomain, TxHash, MaybeExt, Req) ->
+	case get_tx_filename(TxHash) of
+		{ok, _} ->
+			handle_serve_tx_2(ApexDomain, TxHash, MaybeExt, Req);
+		{response, {Status, Headers, Body}} ->
+			{Status, Headers, Body, Req}
+	end.
+
+handle_serve_tx_2(ApexDomain, TxHash, MaybeExt, Req) ->
+	Domain = cowboy_req:host(Req),
+	case get_domain_labeling(ApexDomain, Domain) of
+		apex -> handle_serve_tx_apex_1(ApexDomain, TxHash, Req);
+		{labeled, Label} -> handle_serve_tx_labeled_1(ApexDomain, Label, TxHash, MaybeExt, Req);
+		{custom, Domain} -> handle_serve_tx_apex_1(ApexDomain, TxHash, Req)
+	end.
+
+get_domain_labeling(ApexDomain, Domain) ->
+	case binary:match(Domain, ApexDomain) of
+		{0, _} ->
+			apex;
+		{N, _} ->
+			Label = binary:part(Domain, {0, N-1}),
+			{labeled, Label};
+		nomatch ->
+			get_domain_labeling_1(Domain)
+	end.
+
+get_domain_labeling_1(Domain) ->
+	CustomDomains = ar_meta_db:get(custom_domains),
+	case lists:member(Domain, CustomDomains) of
+		true -> {custom, Domain};
+		false -> error('FIXME')
+	end.
+
+redirect_to_apex(ApexDomain, Req) ->
+	Scheme = cowboy_req:scheme(Req),
+	Port = integer_to_binary(cowboy_req:port(Req)),
+
+	Location =
+		<<Scheme/binary, "://",
+			ApexDomain/binary, ":",
+			Port/binary, "/">>,
+
+	{301, #{<<"location">> => Location}, <<>>, Req}.
+
+resolve_custom_domain_tx(Domain, Req) ->
+	case inet_res:lookup("_arweave." ++ binary_to_list(Domain), in, txt) of
+		[] ->
+			{502, #{}, <<"Domain improperly configured">>, Req};
+		[RecordChunks|_] ->
+			TxHash = list_to_binary(lists:concat(RecordChunks)),
+			handle(<<"GET">>, [<<"tx">>, TxHash, <<"data.html">>], Req)
+	end.
+
+handle_serve_tx_apex_1(ApexDomain, TxHash, Req) ->
+	BlockHash = ?OK(get_tx_block_hash(TxHash)),
+	redirect_to_labeled_tx(ApexDomain, BlockHash, TxHash, Req).
+
+handle_serve_tx_labeled_1(ApexDomain, Label, TxHash, MaybeExt, Req) ->
+	BlockHash = ?OK(get_tx_block_hash(TxHash)),
+	handle_serve_tx_labeled_2(ApexDomain, Label, BlockHash, TxHash, MaybeExt, Req).
+
+get_tx_block_hash(TxHash) ->
+	case ar_tx_search:get_tags_by_id(
+		whereis(http_search_node),
+		ar_util:decode(TxHash),
+		infinity
```

Comment:
Will Cowboy timeout here? If not, we should. I think, makes sense to introduce a dedicated HTTP response, which would essentialy tell the user that the server is busy.

Reply:
So I implemented a handler-wide timeout of 30 seconds here https://github.com/ArweaveTeam/arweave/pull/144/commits/db91056e355911c600b4858d15f2d4447985fe23. I think it still makes sense to pick smaller values for this line and for the status endpoint. What do you think?